core/modules/loader/ServerUdpBase.cc

```diff
@@ -0,0 +1,150 @@
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+// Class header
+#include "loader/ServerUdpBase.h"
+
+// System headers
+#include <iostream>
+#include <unistd.h>
+
+// Third-party headers
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.loader.ServerUdpBase");
+}
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+std::atomic<uint64_t> ServerUdpBase::_msgIdSeq{1};
+
+
+ServerUdpBase::ServerUdpBase(boost::asio::io_service& io_service, std::string const& host, int port)
+    : _ioService(io_service),
+      _socket(io_service, boost::asio::ip::udp::endpoint(boost::asio::ip::udp::v4(), port)),
+      _hostName(host), _port(port) {
+    _receivePrepare(); // Prime the server for an incoming message.
+}
+
+
+void ServerUdpBase::_receiveCallback(boost::system::error_code const& error, size_t bytesRecvd) {
+    _data->advanceWriteCursor(bytesRecvd); // _data needs to know the valid portion of the buffer.
+    if (!error && bytesRecvd > 0) {
+        LOGS(_log, LOG_LVL_INFO, "rCb received(" << bytesRecvd << "):" <<
+                                 ", code=" << error << ", from endpoint=" << _senderEndpoint);
+
+        _sendData = parseMsg(_data, _senderEndpoint);
+        if (_sendData != nullptr) {
+            _sendResponse();
+        } else {
+            _receivePrepare();
+        }
+    } else {
+        LOGS(_log, LOG_LVL_ERROR, "ServerUdpBase::_receiveCallback got empty message, ignoring");
+        _receivePrepare();
+    }
+
+}
+
+
+void ServerUdpBase::_sendResponse() {
+    _socket.async_send_to(boost::asio::buffer(_sendData->getReadCursor(), _sendData->getBytesLeftToRead()),
+                                              _senderEndpoint,
+                          boost::bind(&ServerUdpBase::_sendCallback, this,
+                                      boost::asio::placeholders::error,
+                                      boost::asio::placeholders::bytes_transferred));
+}
+
+
+void ServerUdpBase::sendBufferTo(std::string const& hostName, int port, BufferUdp& sendBuf) {
+    // The socket is not thread safe. To send on "_socket", it needs to be an async send
+    // and then it needs to know when the message was sent so it can return and free the buffer.
+    using namespace boost::asio;
+
+    std::mutex mtx;
+    std::condition_variable cv;
+    bool done = false;
+
+    auto callbackFunc = [&mtx, &cv, &done](const boost::system::error_code& error, std::size_t bytesTransferred) {
+        {
+            std::lock_guard<std::mutex> lock(mtx);
+            done = true;
+        }
+        cv.notify_one();
+    };
+
+    ip::udp::endpoint dest(boost::asio::ip::address::from_string(hostName), port);
+    _socket.async_send_to(buffer(sendBuf.getReadCursor(), sendBuf.getBytesLeftToRead()), dest,
```

Comment:
Why ASYNC I/O if you're going to wait on the CV below?

Reply:
I need to use async io since _socket is not thread safe. Asio will order async writes so they don't cause problems. I'm not sure asio lets you mix synchronous and async use of the socket for this but I was having communication issues, and this appears to work.

Reply:
Hmmm...  Really should be the same as `send_to`...  This extra complication worries me, and makes me wonder if this "fixed" the problem, or just kicked it down the road :thinking:   If you haven't tried `send_to` here instead recently, it might be worth another try in case you addressed the issue you were seeing in the meantime elsewhere?

But if you stick with `async_send_to`, a style suggestion (YMMV): I'd just go ahead and put the lambda directly inline in the call to `async_send_...` in a situation like this.  You are only "using" it once, and nesting it inside the `async_send_to` in my mind makes it more clear when reading the code when it is going to be invoked.